blog/_posts/2020-11-10-introduction-to-network-role.md

```diff
@@ -0,0 +1,298 @@
+---
+layout: post
+title: "Introduction to Network Role"
+section: Blog
+date: 2020-11-10T12:00:00
+author: Wen Liang
+category: talk
+---
+## Introduction
+
+The network role supports two providers: nm and initscripts. For CentOS/RHEL 6,
+we only use initscripts as providers. For CentOS/RHEL 7+, we use initscripts
+and nm as providers. Various networking profiles can be configured via customized
+Ansible module. Several tasks will run for host networking setup, including but
+not limited to, package installation, starting/enabling services. Network role CI
+system  consists of Tox running unit tests and Test-harness running integration
+tests. When we use Tox to run unit tests,  we can check code formatting using
+Python Black, check YAML files formatting etc. Integration tests run in internal
+OpenShift, watch configured GitHub repositories for PRs, check out new PR, run
+all testing playbooks against all configured images, fresh machine for every test
+playbook, sets statuses of PR and uploads results. For better testing efficiency,
+in some playbooks, we can call internal Ansible modules instead of role to skip
+redundant tasks, we can also group Ansible modules into blocks for more targeted
+unit testing. Furthermore, there are helper scripts to get coverage from integration
+tests via Ansible, basic unit for argument parsing, additional helper files for
+assertion/test setup/logging.
+
+## Code structure
+
+The repository is structured as follows:
+- `./defaults/ ` – Contains the default role configuration.
+- `./examples/` – Contains YAML examples for different configurations.
+- `./library/network_connections.py` – Contains the internal Ansible module,
+which is the main script. It controls the communication between the role and
+Ansible, imports the YAML configuration and applies the changes to the provider
+(i.e. NetworkManager, initscripts).
+- `./meta/` – Metadata of the project.
+- `./module_utils/network_lsr/` – Contains other files that are useful for the
+network role (e.g. the YAML argument validator)
+- `./tasks/` – Declaration of the different tasks that the role is going to execute.
+- `./tests/playbooks/` – Contains the complete tests for the role, `./tests/test_*.yml`
+are shims to run tests once for every provider, `./tests/tasks/` contains task
+snippets that are used in multiple tests to avoid having the same code repeated
+multiple times.
+
+
+## How to run test
+
+#### Tox Unit Tests
+- `tox -l`, list all the unit testing, available unit testing options are:
+
+  * black
+  * pylint
+  * flake8
+  * yamllint
+  * py26
+  * py27
+  * py36
+  * py37
+  * py38
+  * collection
+  * custom
+
+- tox, run all the tests
+- tox -e py36, test with Python 3.6, or just tox if you want to run all the tests.
```

Comment:
`py36` is pyunit testing with Python 3.6
omit ", or just tox if you want to run all the tests."